### PR TITLE
Use PiPPy to distribute Mixture of Experts

### PR DESCRIPTION
**Pull request by kwen2501**
Demonstrate an extended use of PiPPy: distribute MoE models along the expert dimension.

```
PiPPy model:
class GraphModule(torch.nn.Module):
    def forward(self, arg14_1: "f32[4, 16]"):
        # No stacktrace found for following nodes
        submod_0 = self.submod_0(arg14_1);  arg14_1 = None
        submod_3 = self.submod_3(submod_0)
        submod_2 = self.submod_2(submod_0)
        submod_1 = self.submod_1(submod_0);  submod_0 = None
        submod_4 = self.submod_4(submod_1, submod_2, submod_3);  submod_1 = submod_2 = submod_3 = None
        return (submod_4,)
        
    class GraphModule(torch.nn.Module):
        def forward(self, arg_0):
            arg14_1: "f32[4, 16]"; 
        
            arg14_1, = fx_pytree.tree_flatten_spec(([arg_0], {}), self._in_spec)
            # No stacktrace found for following nodes
            pre_proc: "f32[4, 16]" = self.pre_proc(arg14_1);  arg14_1 = None
            return pre_proc
            
    class GraphModule(torch.nn.Module):
        def forward(self, addmm: "f32[4, 16]"):
            # No stacktrace found for following nodes
            experts_2: "f32[4, 16]" = getattr(self.experts, "2")(addmm);  addmm = None
            return experts_2
            
    class GraphModule(torch.nn.Module):
        def forward(self, addmm: "f32[4, 16]"):
            # No stacktrace found for following nodes
            experts_1: "f32[4, 16]" = getattr(self.experts, "1")(addmm);  addmm = None
            return experts_1
            
    class GraphModule(torch.nn.Module):
        def forward(self, addmm: "f32[4, 16]"):
            # No stacktrace found for following nodes
            experts_0: "f32[4, 16]" = getattr(self.experts, "0")(addmm);  addmm = None
            return experts_0
            
    class GraphModule(torch.nn.Module):
        def forward(self, addmm_2: "f32[4, 16]", addmm_4: "f32[4, 16]", addmm_6: "f32[4, 16]"):
            # File: /data/users/kw2501/PiPPy/examples/MoE/dist_moe.py:47 in forward, code: return torch.cat(outputs, dim=1)
            cat_default: "f32[4, 48]" = torch.ops.aten.cat.default([addmm_2, addmm_4, addmm_6], 1);  addmm_2 = addmm_4 = addmm_6 = None
            return cat_default
            
PiPPy model equivalent: 2.4520821571350098 ref 2.4520821571350098
NCCL version 2.20.5+cuda12.4
Distributed model equivalent: 2.4520821571350098 ref 2.4520821571350098
dist_out: torch.Size([4, 48])
ref_out: torch.Size([4, 48])
```

### CODE REVIEW

File: examples/mixture_of_experts/dist_moe.py [hunk 1]
```diff
@@ -0,0 +1,96 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates
+# torchrun --nproc-per-node 5 dist_moe.py
```
Comment (lessw2020) [inline examples/mixture_of_experts/dist_moe.py:2]: I think we should keep the standard format and preface the run instructions as usual:
"# Minimum effort to run this example:"
I actually missed this line b/c of it and had to work backwards until I figured out it was expecting n_experts+2 for world size.

File: examples/mixture_of_experts/dist_moe.py [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+        model, {f"experts.{i}": SplitPoint.END}
+    )
+
+pippy_model = pipeline(model, 1, (x,))
+
+assert pippy_model.num_stages == world_size
+if rank == 0:
+    print("Original model:\n", model)
+    print("PiPPy model:")
+    pippy_model.print_readable()
```
Comment (lessw2020) [inline examples/mixture_of_experts/dist_moe.py:72]: I got an error on this line.  Added a debug print for type:
`type <class 'pippy.IR.Pipe'>
PiPPy model equivalent: 2.4520821571350098 ref 2.4520821571350098
[rank0]: Traceback (most recent call last):
[rank0]:   File "/data/users/less/local/pippy/examples/moe/moe.py", line 74, in <module>
[rank0]:     pippy_model.print_readable()
[rank0]:   File "/home/less/local/miniconda3/envs/gocuda/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1709, in __getattr__
[rank0]:     raise AttributeError(f"'{type(self).__name__}' object has no attribute '{name}'")
[rank0]: AttributeError: 'Pipe' object has no attribute 'print_readable'`

File: examples/mixture_of_experts/dist_moe.py [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+
+# Each expert is a MLP
+class ExpertLayer(torch.nn.Module):
+    def __init__(self, d_hid):
+        super(ExpertLayer, self).__init__()
+        self.net1 = torch.nn.Linear(d_hid, d_hid)
+        self.relu = torch.nn.ReLU()
+        self.net2 = torch.nn.Linear(d_hid, d_hid)
+
+    def forward(self, x):
```
Comment (lessw2020) [inline examples/mixture_of_experts/dist_moe.py:25]: nit for consistency - return type:
`-> torch.Tensor:`

File: examples/mixture_of_experts/dist_moe.py [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+
+d_hid = 16
+n_experts = 3
+batch_size = 4
+
+torch.manual_seed(0)
+
+# Each expert is a MLP
+class ExpertLayer(torch.nn.Module):
+    def __init__(self, d_hid):
```
Comment (lessw2020) [inline examples/mixture_of_experts/dist_moe.py:19]: could add return type here to match the MoE class init format:
`-> None:`

**Review by lessw2020 — APPROVED** [review]
this is a fantastic concept implementation! 
I added a few nits on typing, suggestion re: noting min effort. 
I hit an error re: print_readable for the pippy model (see comment with details) so I had to comment that out to complete.
Otherwise looks great and really nice to see MoE being added here! 
-------------- run results below --------------------
[less@devgpu003.cco3 /data/users/less/local/pippy/examples/moe (main)]$ torchrun --nproc-per-node 5 moe.py
...
PiPPy model equivalent: 2.4520821571350098 ref 2.4520821571350098
PiPPy model equivalent: 2.4520821571350098 ref 2.4520821571350098
Original model:
 MoE(
  (pre_proc): Linear(in_features=16, out_features=16, bias=True)
  (experts): ModuleList(
    (0-2): 3 x ExpertLayer(
      (net1): Linear(in_features=16, out_features=16, bias=True)
      (relu): ReLU()
      (net2): Linear(in_features=16, out_features=16, bias=True)
    )
  )
)
PiPPy model:
PiPPy model equivalent: 2.4520821571350098 ref 2.4520821571350098
PiPPy model equivalent: 2.4520821571350098 ref 2.4520821571350098
PiPPy model equivalent: 2.4520821571350098 ref 2.4520821571350098
NCCL version 2.20.5+cuda12.4
Distributed model equivalent: 2.4520816802978516 ref 2.4520821571350098
dist_out: torch.Size([4, 48])
ref_out: torch.Size([4, 48])

File: examples/mixture_of_experts/dist_moe.py (hunk 2 above)
Comment (kwen2501) [reply examples/mixture_of_experts/dist_moe.py:72]: Sorry, the `print_readable()` method is in a pending PR: https://github.com/pytorch/PiPPy/pull/1037

**Pull request merged**